design/architecture/api.md

```diff
@@ -1,22 +1,285 @@
-well-documented would ensure the linked data would be FAIR, safely and
-securely.
-
-In this case, the API would be between the user and the web server that
-stores the underlying database and documentation. The API would be a
-combination of a predefined set of instructions that are sent to the web
-server to run certain commands as well as a set of explicit conventions
-and rules on how files and folders are structured and named. Taken
-together, this API would allow other software like R packages to be
-built to interact with the backend to automate tasks done by the users.
-
-TODO: File tree structure design with API endpoints.
+## API security
+
+Adding a security layer to an API is crucial for ensuring the
+confidentiality, integrity, and availability of the data and systems
+that use the API. APIs are often used to connect different systems and
+applications, and they provide a way for external parties to access and
+interact with data. Because of this, they can be an attractive target
+for cyber attackers looking to gain unauthorized access, steal sensitive
+information, or disrupt service availability. To protect against these
+threats, it is important to implement a robust security strategy for the
+API.
+
+### Approach 1: Basic Authentication
+
+Basic Authentication is a simple authentication scheme used by APIs to
+authenticate users based on a username and password. In this scheme, the user's
+credentials are encoded in a Base64-encoded string and included in the HTTP
+Authorization header of each request.The server verifies the credentials and
+either grants or denies access to the requested resource.
+
+### Approach 2: User Token for API requests
+
+User Token is a type of authentication token used by APIs to authenticate and
+authorize user requests. When a user logs in, the API generates a unique token
+for that user, which is then stored securely on the client side (seedcase box).
+User Tokens are more secure than Basic Authentication because they are not
+transmitted in plaintext and can be set to expire after a certain period of
+time, forcing the user to log in again to obtain a new token. They can also be
+revoked by the server if the user's access needs to be terminated.
+
+### Approach 3: OAuth
+
+Use OAuth approach is more complicated compared to previously two. Here
+are some basic steps.
+
+1. User request access to the project data of the seedcase.
+2. Seedcase redirect the OAuth server (e.g. Shib-Identity-Provider)
+3. OAuth Server authenticates user by username and password.
+4. OAuth Server sends secret key back the seedcase
+5. Seedcase used the secret key from OAuth server provide correct
+   authorization of the user for accessing the data
+
+## API Endpoint: POST /projects/{project_id}/research-data
+
+This endpoint allows users to upload raw research data to a project.
+
+#### Request parameters
+
+project_id (string, required) - The unique identifier of the project that the
+data is being posted to.
```

Comment:
Just reviewing this right now. I think one thing to clear up is that each Seedcase "instance" would contain data for only one "project"/study. For instance, one Seedcase instance would only contain data from DD2. If someone wants to use seedcase for another study7data source, they would need to create another "instance". So we don't need a `project_id` for data input. We would have a `project_id` for the requests to use the data (a "data project" / "data extract"). I will add more detail to that in the introduction document.